zp-relayer/direct-deposit/BatchCache.ts

```diff
@@ -0,0 +1,69 @@
+import type Redis from 'ioredis'
+
+export class BatchCache<T> {
+  private timer: NodeJS.Timeout | null = null
+
+  constructor(
+    private batchSize: number,
+    private ttl: number,
+    private cb: (es: T[]) => Promise<void> | void,
+    private redis: Redis,
+    private key: string = 'dd:cache'
+  ) {}
+
+  private clearTimer() {
+    if (this.timer) {
+      clearTimeout(this.timer)
+    }
+  }
+
+  private setTimer() {
+    this.clearTimer()
+    this.timer = setTimeout(() => this.execute(), this.ttl)
+  }
+
+  private addToRedis(values: [number, T][]) {
+    const rawValues = values.map(([s, v]) => [s, JSON.stringify(v)] as [number, string])
+    return this.redis.zadd(this.key, ...rawValues.flat())
+  }
+
+  private async take(count: number) {
+    const es: string[] = await this.redis.zpopmin(this.key, count)
+    return es.map(e => JSON.parse(e))
+  }
+
+  // TODO: count could be maintained in-memory
+  private count() {
+    return this.redis.zcount(this.key, '-inf', '+inf')
```

Comment:
There is `ZCARD` - https://redis.io/commands/zcard/

Reply:
Oh, thanks

Reply:
Updated